### ISSUE TITLE
Construct 3 Android Export issue 

### ISSUE THREAD

**Issue by LucaDiMeglio** [Scirra/Construct-3-bugs#3090]
When I try to export my android game with phonegap i get an error and in the log the error is this:
:app:compileDebugNdk NO-SOURCE
:app:compileDebugSources
:CordovaLib:mergeDebugShaders
:CordovaLib:compileDebugShaders
:CordovaLib:generateDebugAssets
:CordovaLib:mergeDebugAssets
:app:mergeDebugShaders
:app:compileDebugShaders
:app:generateDebugAssets
:app:mergeDebugAssets
:app:transformClassesWithStackFramesFixerForDebug
:app:transformClassesWithDesugarForDebug
:app:transformClassesWithDexBuilderForDebug
:app:transformDexArchiveWithExternalLibsDexMergerForDebug FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:transformDexArchiveWithExternalLibsDexMergerForDebug'.
> java.lang.RuntimeException: java.lang.RuntimeException: com.android.builder.dexing.DexArchiveMergerException: Unable to merge dex

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

I have this issue from about 2 or 3 weeks and the problem is not the new code that i added after because i tried to export again an old version of my game and i get the same error message in the log. I use android export 5.0+ lollipop- cordova project.

## Observed result
I can't export my game

I get an error
## Expected result
I should be able to export the game

i think it as has been the r 148, but probably the version is not the problem beacause i am pretty sure that once i exported with r148 without having problem, i tried with newest version too.

**Comment by shortercode** [comment]
This would appear to be related to Android X issues which we've talked about on other bug reports. If you can find any reference to `android.support` or `androidx.` in your error messages then that will tell you, but I can't remember how much information PGB gives you.

Android X is a replacement for a commonly used component in the Android exosystem. If any part of your project uses Android X then you can't use the old component anywhere else. 

On June 17th the teams behind the google play services SDKs updated them all to use Android X. At the time this broke our Google Play plugin, as it uses the latest version of the google play services games SDK. We modified our cordova plugin to use the older version so that it wouldn't cause compilation issues for older release of construct, the release being `construct-play-games@1.0.21`.

One of our aims is to keep up to date with google play services so in r156 we updated both the Google Play plugin and the Mobile Advert plugin to use the new Android X versions. The associated cordova plugins revisions should not be being used with older versions of C3.

I wouldn't expect the new versions to work on PGB, as compiling with Android X support on cordova currently requires the use of a hook script which they do not support. I don't believe the cordova team has an answer to Android X properly yet, and the PGB team tend be a fair bit behind them. So I wouldn't expect this to be resolved by them anytime soon.

To summarise: if you use specifically `construct-play-games@1.0.21` and `construct-mobile-advert@1.2.12`  then you should be fine for the time being. But phonegap build doesn't work with Android X, and moving forwards we're intending to support it. 

You may also have problems with other cordova plugins if your using them, you should talk to the maintainers if that is the case.

**Comment by LucaDiMeglio** [comment]
I tried to do an export in the r156.2 directly from construct with an unsigned apk: i got this message maybe it is the thign that you talked about? 
**Error: Discovered platform "android@^8.0.0" in config.xml or package.json. Adding it to the project
Using cordova-fetch for cordova-android@^8.0.0
Adding android project...
Creating Cordova project for the Android platform:
Path: ~~/android


Activity: MainActivity
Android target: android-28
Subproject Path: CordovaLib
Subproject Path: app
Android project created with cordova-android@8.0.0
Plugin 'construct-play-games' found in config.xml... Migrating it to package.json
Plugin 'cordova-plugin-inappbrowser' found in config.xml... Migrating it to package.json
Plugin 'construct-mobile-advert' found in config.xml... Migrating it to package.json
Plugin 'cordova-plugin-whitelist' found in config.xml... Migrating it to package.json
Plugin 'cordova-plugin-file' found in config.xml... Migrating it to package.json
Plugin 'cordova-plugin-fullscreen' found in config.xml... Migrating it to package.json
Plugin 'construct-mobile-notchfix' found in config.xml... Migrating it to package.json
Discovered saved plugin "construct-play-games". Adding it to the project
Installing "construct-play-games" for android
Subproject Path: CordovaLib
Subproject Path: app
Discovered saved plugin "cordova-plugin-inappbrowser". Adding it to the project
Installing "cordova-plugin-inappbrowser" for android
Discovered saved plugin "construct-mobile-advert". Adding it to the project
Failed to restore plugin "construct-mobile-advert" from config.xml. You might need to try adding it again. Error: Variable(s) missing (use: --variable ADMOB_APP_ID=value).
Discovered saved plugin "cordova-plugin-whitelist". Adding it to the project
Installing "cordova-plugin-whitelist" for android

This plugin is only applicable for versions of cordova-android greater than 4.0. If you have a previous platform version, you do *not* need this plugin since the whitelist will be built in.

Discovered saved plugin "cordova-plugin-file". Adding it to the project
Installing "cordova-plugin-file" for android
Installing "cordova-plugin-compat" for android
Plugin doesn't support this project's cordova-android version. cordova-android: 8.0.0, failed version requirement: 
<6.3.0
Skipping 'cordova-plugin-compat' for android

The Android Persistent storage location now defaults to "Internal". Please check this plugin's README to see if your application needs any changes in its config.xml.

If this is a new application no changes are required.

If this is an update to an existing application that did not specify an "AndroidPersistentFileLocation" you may need to add:

"<preference name="AndroidPersistentFileLocation" value="Compatibility" />"

to config.xml in order for the application to find previously stored files.

Discovered saved plugin "cordova-plugin-fullscreen". Adding it to the project
Installing "cordova-plugin-fullscreen" for android
Discovered saved plugin "construct-mobile-notchfix". Adding it to the project
Installing "construct-mobile-notchfix" for android**

**Comment by shortercode** [comment]
That's a slightly different issue. The latest version of the Admob SDK
requires the application ID be specified at compile time. So we've added a
Cordova variable for it and tied it to the plugins android app ID property.
Unfortunately we don't have a user friendly way to mention when the value
is absent at the moment, so the build just fails.

It's also worth noting that if the app ID is invalid the Admob SDK will
crash your app when it starts for "security reasons".

**Comment by LucaDiMeglio** [comment]
You are right thank you for your help :+1: .
